### PR TITLE
feat(memory): add deduplication detection service with CLI command (#160)

### LINKED ISSUE #160: feat(memory): Add deduplication detection service with CLI command
## Problem

As Lisa accumulates facts over weeks and months, duplicates appear — the same decision recorded in different sessions, similar facts captured from different prompts, or manually added facts that overlap with auto-captured ones. There is no way to detect these duplicates. The nate-plan (Principle 6.2) calls for deduplication and consolidation. This issue implements detection; consolidation (the action on detected duplicates) follows in 5C.

## Proposed Solution

Create a `DeduplicationService` with a three-pass detection algorithm (exact match → tag overlap → Jaccard similarity) and a `lisa memory dedupe` CLI command for detection-only output.

| File | Action | Details |
|------|--------|---------|
| `src/lib/domain/interfaces/IDeduplicationService.ts` | Create | `IDeduplicationService`, `IDuplicateGroup`, `IDeduplicationResult` interfaces |
| `src/lib/infrastructure/services/DeduplicationService.ts` | Create | Three-pass detection: exact match, tag overlap, Jaccard word similarity |
| `src/lib/infrastructure/di/tokens.ts` | Modify | Add `DeduplicationService` token |
| `src/lib/infrastructure/di/bootstrap.ts` | Modify | Register `DeduplicationService` with injected `MemoryService` |
| `src/lib/domain/interfaces/index.ts` | Modify | Export new deduplication interfaces |
| `src/lib/commands/knowledge.ts` | Modify | Add `lisa memory dedupe` command with `--min-similarity`, `--limit`, `--since` options |
| `src/project/.lisa/skills/memory/memory.ts` | Modify | Add `dedupe` subcommand handler |
| `tests/unit/.../services/DeduplicationService.test.ts` | Create | Tests for all three detection passes, similarity calculation, edge cases |

### Detection Algorithm

1. **Pass 1 — Exact match**: Group facts with identical normalized text (case-insensitive, trimmed, collapsed whitespace). `similarity: 1.0`, `reason: 'exact-match'`
2. **Pass 2 — Tag overlap**: Group facts sharing the same `type:*` tag where text shares 3+ common words (after removing stop words). `similarity: 0.5–0.8`, `reason: 'tag-overlap'`
3. **Pass 3 — Similar content**: Jaccard similarity on word sets between facts within the same category. Threshold: `minSimilarity` (default 0.6). `reason: 'similar-content'`

A fact appears in at most one group (highest-similarity group wins). Semantic similarity via embeddings deferred to Phase 6.

## Acceptance Criteria

- [ ] Detects exact duplicate facts (case-insensitive, whitespace-normalized)
- [ ] Detects tag-overlap duplicates (same type tag, shared words)
- [ ] Detects content-similar facts using Jaccard similarity
- [ ] `lisa memory dedupe` displays detected groups with similarity scores
- [ ] Results are detection-only — no mutations (consolidation is in 5C)
- [ ] Similarity threshold is configurable via `--min-similarity`
- [ ] New code is covered by unit tests
- [ ] All tests pass (`npm test`)
- [ ] Linting passes (`npm run lint`)
- [ ] TypeScript compiles (`npx tsc --noEmit`)

## References

- Part of #158
- `.dev/features/nate-plan.md` (Principle 6.2: Deduplication and Consolidation)
- Semantic similarity (embeddings) deferred to Phase 6

### PR DESCRIPTION
**Pull request by TonyCasey**
## Summary

Implements the deduplication detection service (Phase 5B of the nate-plan) with a three-pass algorithm for identifying duplicate memory facts.

**Changes:**
- **Domain layer**: New `IDeduplicationService` interface with `IDuplicateGroup`, `IDeduplicationResult`, `IDeduplicationOptions` types
- **Infrastructure layer**: `DeduplicationService` with three-pass algorithm:
  - Pass 1: Exact match on normalized text (case-insensitive, whitespace-collapsed)
  - Pass 2: Tag overlap via conflict groups + 3+ shared meaningful words
  - Pass 3: Jaccard word-set similarity with configurable threshold (default 0.6)
- **DI registration**: Registered as transient service in bootstrap
- **Skill-level integration**: `MemoryService.dedupe()` loads facts from Neo4j and runs algorithm
- **CLI command**: `lisa memory dedupe` with `--min-similarity`, `--limit`, `--since`, `--group` options
- **Pure exported functions**: `normalizeText`, `extractWords`, `jaccardSimilarity`, `detectDuplicatesFromFacts` for reuse

**Testing:**
- 42 new unit tests (36 for DeduplicationService, 6 for MemoryCliService dedupe command)
- All 1107 tests pass

Closes #160

## Test plan

- [x] Unit tests for `normalizeText()` (lowercase, trim, whitespace collapse, tabs/newlines)
- [x] Unit tests for `extractWords()` (meaningful words, stop words, punctuation, case-insensitive)
- [x] Unit tests for `jaccardSimilarity()` (identical, disjoint, partial overlap, empty sets, subset)
- [x] Unit tests for `detectDuplicatesFromFacts()` Pass 1 (exact match, case-insensitive, whitespace-normalized, 3+ members)
- [x] Unit tests for `detectDuplicatesFromFacts()` Pass 2 (tag overlap with 3+ words, fewer than 3 words, skip claimed)
- [x] Unit tests for `detectDuplicatesFromFacts()` Pass 3 (above threshold, below threshold, transitive groups)
- [x] Unit tests for edge cases (limit, empty facts, single fact, no uuid, empty text, sort order, one-group guarantee)
- [x] Unit tests for MemoryCliService dedupe command (default options, minSimilarity, limit, since, invalid since, explicit group)
- [x] Full test suite passes (1107 tests)
- [x] Lint clean (0 errors, 0 warnings)
- [x] Build clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Adds `lisa memory dedupe` CLI to detect duplicate facts (no automatic changes), with --group, --min-similarity, --limit, --since options; returns grouped duplicate results and counts
  * Three-pass detection: exact text match, tag overlap, and content similarity; configurable sensitivity and result limiting

* **Tests**
  * Comprehensive unit tests validating dedupe logic, helpers, thresholds, and edge cases

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 ### Added
 
+#### Deduplication Detection Service with CLI Command ([#160](https://github.com/TonyCasey/lisa/issues/160))
+
+Added a three-pass deduplication detection service for finding duplicate memory facts without mutating them.
+
+- `DeduplicationService` with three-pass algorithm: exact match, tag overlap, Jaccard word-set similarity
+- `IDeduplicationService`, `IDuplicateGroup`, `IDeduplicationResult` domain interfaces
+- `lisa memory dedupe` CLI command with `--min-similarity`, `--limit`, `--since` options
+- Skill-level `dedupe` command in `MemoryCliService` and `MemoryService`
```
[the remark this answers is not in the document]
Comment (TonyCasey) [reply CHANGELOG.md]: Fixed — added `--group` to the CHANGELOG entry.

File: src/lib/infrastructure/services/DeduplicationService.ts [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+          if (b.words.has(w)) commonCount++;
+        }
+        if (commonCount >= 3) {
+          if (!tagGroup.includes(a.fact)) tagGroup.push(a.fact);
+          if (!tagGroup.includes(b.fact)) tagGroup.push(b.fact);
+          const ratio = jaccardSimilarity(a.words, b.words);
+          if (ratio > maxOverlapRatio) maxOverlapRatio = ratio;
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (TonyCasey) [reply src/lib/infrastructure/services/DeduplicationService.ts]: Fixed — replaced `Array.includes()` with a `Set<string>` for O(1) membership checks using uuid-based tracking.

File: src/lib/skills/shared/services/MemoryService.ts [hunk 3]
```diff
@@ -393,5 +395,117 @@ export function createMemoryService(deps: IMemoryServiceDependencies): IMemorySe
...
+
+        return {
+          status: 'ok',
+          action: 'dedupe',
+          group: groupId,
+          totalFactsScanned: facts.length,
+          duplicateGroups: skillGroups,
+          totalDuplicates: skillGroups.length,
+          minSimilarity,
+          mode: 'neo4j',
```
[the remark this answers is not in the document]
Comment (TonyCasey) [reply src/lib/skills/shared/services/MemoryService.ts]: Fixed — `totalDuplicates` now sums `group.facts.length` across all groups using `reduce()`, matching the infrastructure-level service behavior. Updated in both the infrastructure `DeduplicationService` and the skill-level `MemoryService`.

File: tests/unit/src/lib/infrastructure/services/DeduplicationService.test.ts [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+      it('should handle default options', () => {
+        const facts = [
+          makeFact({ uuid: 'a1', fact: 'same text' }),
+          makeFact({ uuid: 'a2', fact: 'same text' }),
+        ];
+        // No options passed — should use defaults
+        const result = detectDuplicatesFromFacts(facts, []);
+        assert.strictEqual(result.length, 1);
+      });
+    });
```
[the remark this answers is not in the document]
Comment (TonyCasey) [reply tests/unit/src/lib/infrastructure/services/DeduplicationService.test.ts]: Acknowledged — however, the existing test suite across the entire codebase consistently uses the \ naming pattern (e.g., all 1107 tests). Renaming just these tests would create inconsistency. We'll consider migrating to the AAA naming convention as a codebase-wide effort in a separate task.

File: tests/unit/src/lib/skills/shared/services/MemoryCliService.test.ts [hunk 5]
```diff
@@ -368,11 +385,81 @@ describe('MemoryCliService', () => {
...
+    it('should use explicit group if provided', async () => {
+      await cliService.run({
+        ...defaultArgs(),
+        command: 'dedupe',
+        explicitGroup: 'custom-group',
+      });
+
+      assert.strictEqual(dedupeCalls[0].groupId, 'custom-group');
+    });
+  });
```
[the remark this answers is not in the document]
Comment (TonyCasey) [reply tests/unit/src/lib/skills/shared/services/MemoryCliService.test.ts]: Same as above — deferring the AAA naming convention migration to a codebase-wide task to maintain consistency with the existing 1107 tests.

File: tests/unit/src/lib/infrastructure/services/DeduplicationService.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (TonyCasey) [reply tests/unit/src/lib/infrastructure/services/DeduplicationService.test.ts]: Acknowledged. The existing test suite across the codebase (1107 tests) consistently uses the 'should ...' naming pattern. Renaming just these tests would create inconsistency. We will consider migrating to the AAA naming convention as a codebase-wide effort in a separate task.
